File: assign1/PHP/src/Name.php
<?php

function imperativeAverageNumberOfLetters(array $names) {
  if (count($names) == 0) {
    return 0.0;
  }

  $sum = 0;
	
  foreach ($names as $name) {
    $sum = $sum + strlen($name);
  }
	
  return $sum / count($names);	    
}

function imperativeGroupNamesByFirstLetter(array $names) {
  $groupedNames = [];

  foreach($names as $name) {
    $firstLetter = substr($name, 0, 1);

    if (array_key_exists($firstLetter, $groupedNames)) {
      array_push($groupedNames[$firstLetter], $name);
    } else {
      $groupedNames[$firstLetter] = array($name);
    }
  }
  return $groupedNames;
}

function imperativeCountNamesByFirstLetter(array $names) {
  $countedNames = [];

  foreach($names as $name) {
    $firstLetter = substr($name, 0, 1);

    if (array_key_exists($firstLetter, $countedNames)) {
      $countedNames[$firstLetter]++;
    } else {
      $countedNames[$firstLetter] = 1;
    }
  }
  return $countedNames;
}

function functionalAverageNumberOfLetters(array $names){
  if (empty($names)){
    return 0.0;
  }

  return array_sum(array_map(function($name) { return strlen($name); }, $names)) / count($names);
    
}

function functionalGroupNamesByFirstLetter(array $names){
  if (empty($names)){
    return [];
  }
    
  return array_map(function($nameList) {
    return (array) $nameList;
    }, array_merge_recursive(...array_map(function($name) {
      return array(substr($name, 0, 1) => $name);
    }, $names)));
}

function functionalCountNamesByFirstLetter(array $names){
  if (empty($names)){
    return [];
  }
  
  return array_count_values(array_map(function($name) { return(substr($name, 0, 1)); }, $names));
}


function givenList() {
  $names = ["Fadi", "Timothy", "Derrick", "Omar", "Hiep", "Alexander", "Daniel", "Isela", 
    "Jackqueline", "Bryan", "Nguyen", "Maxim", "Edgar", "Ahson", "Abdulkadir", "En-Kai", "Karla", 
    "Panpan", "You", "Andrew", "Trevor", "Hoang", "Johnson", "Janelle", "Moujie", "MD", "Albert", 
    "Jiahui", "Alejandro", "Venkat"];

  echo "GIVEN LIST\n";
  echo "IMPERATIVE IMPLEMENTATION\n";	
  echo "Average Number of Letters: \n";
  echo imperativeAverageNumberOfLetters($names) . "\n";
  echo "Group Names By First Letter: \n";
  print_r(imperativeGroupNamesByFirstLetter($names));
  echo "\nCount Names By First Letter: \n";
  print_r(imperativeCountNamesByFirstLetter($names));
  echo "\n";
  
  echo "FUNCTIONAL IMPLEMENTATION\n";
  echo "Average Number of Letters: \n"; 
  echo functionalAverageNumberOfLetters($names) . "\n";
  echo "Group Names By First Letter: \n";
  print_r(functionalGroupNamesByFirstLetter($names));
  echo "\nCount Names By First Letter: \n";
  print_r(functionalCountNamesByFirstLetter($names));
  echo "\n";
}

?>

File: assign1/Scala/src/main/scala/Name.scala
package assign1

import scala.collection.mutable

object Name {
  def imperativeAverageNumberOfLetters(names : List[String]) : Double = {
    if (names.isEmpty) {
      0.0
    } else {
      var sum = 0.0

      for (name <- names){
        sum += name.length()
      }

      sum / names.length
    }
  }


  def imperativeGroupNamesByFirstLetter(names: List[String]): mutable.Map[Char, List[String]] = {
    val groupedNames = mutable.Map[Char, List[String]]()

    for (name <- names){
      val firstCharacter = name(0)

      if (!groupedNames.contains(firstCharacter)) {
        groupedNames(firstCharacter) = List()
      }

      groupedNames(firstCharacter) = groupedNames(firstCharacter) ::: List(name)
    }

    groupedNames
  }

  def imperativeCountNamesStartingWithEachLetter(names : List[String]): mutable.Map[Char, Int] ={
    val countedNames = mutable.Map[Char, Int]()

    for (name <- names){
      val firstCharacter = name(0)

      if (!countedNames.contains(firstCharacter)) {
        countedNames.addOne(firstCharacter -> 0)
      }

      countedNames(firstCharacter) = countedNames(firstCharacter) + 1
    }

    countedNames
  }

  def functionalAverageNumberOfLettersInName(names : List[String]) : Double = {
    if (names.isEmpty){
      0.0
    }else{
      names.map(_.length).sum.toDouble / names.length.toDouble
    }
  }
  
  def functionalGroupNamesByFirstLetter(names : List[String]) : Map[Char, List[String]] = {
    if (names.isEmpty) {
      Map.empty[Char, List[String]]
    } else {
      names.groupBy(_(0))
    }
  }

  def functionalCountNamesStartingWithEachLetter(names : List[String]) : Map[Char, Int] = {
    if (names.isEmpty) {
      Map.empty[Char, Int]
    } else {
      names.groupBy(_(0)).map { case(key, value) => key -> value.size }
    }
  }

  def main(args : Array[String]) : Unit = {
    val names = List("Fadi", "Timothy", "Derrick", "Omar", "Hiep", "Alexander", "Daniel", "Isela", "Jackqueline", "Bryan", "Nguyen", "Maxim", "Edgar", "Ahson", "Abdulkadir", "En-Kai", "Karla", "Panpan", "You", "Andrew", "Trevor", "Hoang", "Johnson", "Janelle", "Moujie", "MD", "Albert", "Jiahui", "Alejandro", "Venkat")

    println("Imperative implementation")
    println("Average number of letters in name: " + imperativeAverageNumberOfLetters(names))
    println("Group names by first letter: " + imperativeGroupNamesByFirstLetter(names))
    println("Count names starting with each letter: " + imperativeCountNamesStartingWithEachLetter(names))
    
    println("Functional implementation")
    println("Average number of letters in name: " + functionalAverageNumberOfLettersInName(names))
    println("Group names by first letter: " + functionalGroupNamesByFirstLetter(names))
    println("Count names starting with each letter: " + functionalCountNamesStartingWithEachLetter(names))
  }
}
